## Setup

In [1]:
import pandas as pd

data_path = "/home/kyle/Documents/HIGGS.csv"

N = 1500000 #1,500,000 to make reading easier
data = pd.read_csv(data_path,nrows=N,header=None)
test_data = pd.read_csv(data_path,nrows=500000,header=None,skiprows=N)
data.head()

0         1         2         3         4         5         6         7   \
0  1.0  0.869293 -0.635082  0.225690  0.327470 -0.689993  0.754202 -0.248573   
1  1.0  0.907542  0.329147  0.359412  1.497970 -0.313010  1.095531 -0.557525   
2  1.0  0.798835  1.470639 -1.635975  0.453773  0.425629  1.104875  1.282322   
3  0.0  1.344385 -0.876626  0.935913  1.992050  0.882454  1.786066 -1.646778   
4  1.0  1.105009  0.321356  1.522401  0.882808 -1.205349  0.681466 -1.070464   

         8         9     ...           19        20        21        22  \
0 -1.092064  0.000000    ...    -0.010455 -0.045767  3.101961  1.353760   
1 -1.588230  2.173076    ...    -1.138930 -0.000819  0.000000  0.302220   
2  1.381664  0.000000    ...     1.128848  0.900461  0.000000  0.909753   
3 -0.942383  0.000000    ...    -0.678379 -1.360356  0.000000  0.946652   
4 -0.921871  0.000000    ...    -0.373566  0.113041  0.000000  0.755856   

         23        24        25        26        27        28  
0  0.979563  0.978076  0.920005  0.721657  0.988751  0.876678  
1  0.833048  0.985700  0.978098  0.779732  0.992356  0.798343  
2  1.108330  0.985692  0.951331  0.803252  0.865924  0.780118  
3  1.028704  0.998656  0.728281  0.869200  1.026736  0.957904  
4  1.361057  0.986610  0.838085  1.133295  0.872245  0.808487  

[5 rows x 29 columns]

In [2]:
import numpy as np
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation
from keras.optimizers import SGD, RMSprop, Adadelta, Adam
from sklearn.metrics import roc_auc_score

y = np.array(data.loc[:,0])
x = np.array(data.loc[:,1:])
x_test = np.array(test_data.loc[:,1:])
y_test = np.array(test_data.loc[:,0])


Using TensorFlow backend.


In [56]:
data[0].value_counts()

1.0    793773
0.0    706227
Name: 0, dtype: int64

In [58]:
total = 793773 + 706227
pos_rate = 793773 / total
pos_rate

0.529182

## Basic Model Supplied with the Assignment

In [15]:
model = Sequential()

model.add(Dense(50, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 28 here
model.add(Activation('sigmoid'))
model.add(Dropout(0.10))

model.add(Dense(50, kernel_initializer='uniform'))
model.add(Activation('sigmoid'))
model.add(Dropout(0.10))

model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test,model.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 10s 6us/step - loss: 0.6919 - acc: 0.5273
Epoch 2/5
1500000/1500000 [==============================] - 9s 6us/step - loss: 0.6918 - acc: 0.5279
Epoch 3/5
1500000/1500000 [==============================] - 9s 6us/step - loss: 0.6786 - acc: 0.5600
Epoch 4/5
1500000/1500000 [==============================] - 9s 6us/step - loss: 0.6496 - acc: 0.6147
Epoch 5/5
1500000/1500000 [==============================] - 9s 6us/step - loss: 0.6405 - acc: 0.6246


0.6898656691309724

## Work

### Part 1 - Pick 3 or more different architectures (add/subtract layers+neurons) and run the model + score. 

#### Model 1
Five hidden layers (50 neurons each) 10 percent drop out at each hidden layer. No improvement.

In [30]:
five_layer_model = Sequential()
# Layer 1
five_layer_model.add(Dense(50, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 28 here
five_layer_model.add(Activation('sigmoid'))
five_layer_model.add(Dropout(0.10))

# Layer 2
five_layer_model.add(Dense(50, kernel_initializer='uniform'))
five_layer_model.add(Activation('sigmoid'))
five_layer_model.add(Dropout(0.10))

# Layer 3
five_layer_model.add(Dense(50, kernel_initializer='uniform')) 
five_layer_model.add(Activation('sigmoid'))
five_layer_model.add(Dropout(0.10))

# Layer 4
five_layer_model.add(Dense(50, kernel_initializer='uniform')) 
five_layer_model.add(Activation('sigmoid'))
five_layer_model.add(Dropout(0.10))

# Layer 5
five_layer_model.add(Dense(50, kernel_initializer='uniform')) 
five_layer_model.add(Activation('sigmoid'))
five_layer_model.add(Dropout(0.10))

# Output
five_layer_model.add(Dense(1, kernel_initializer='uniform')) 
five_layer_model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
five_layer_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

five_layer_model.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test,five_layer_model.predict(x_test))


Epoch 1/5
1500000/1500000 [==============================] - 16s 11us/step - loss: 0.6919 - acc: 0.5277
Epoch 2/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.6918 - acc: 0.5280
Epoch 3/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.6918 - acc: 0.5282
Epoch 4/5
1500000/1500000 [==============================] - 16s 10us/step - loss: 0.6917 - acc: 0.5286
Epoch 5/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.6917 - acc: 0.5288


0.5016195849252397

#### Model 2
Increasing the neurons in each layer to 100 to see what happens

In [31]:
# Model 2
five_layer_model = Sequential()

# Layer 1
five_layer_model.add(Dense(100, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 28 here
five_layer_model.add(Activation('sigmoid'))
five_layer_model.add(Dropout(0.10))

# Layer 2
five_layer_model.add(Dense(100, kernel_initializer='uniform'))
five_layer_model.add(Activation('sigmoid'))
five_layer_model.add(Dropout(0.10))

# Layer 3
five_layer_model.add(Dense(100, kernel_initializer='uniform')) 
five_layer_model.add(Activation('sigmoid'))
five_layer_model.add(Dropout(0.10))

# Layer 4
five_layer_model.add(Dense(100, kernel_initializer='uniform')) 
five_layer_model.add(Activation('sigmoid'))
five_layer_model.add(Dropout(0.10))

# Layer 5
five_layer_model.add(Dense(100, kernel_initializer='uniform')) 
five_layer_model.add(Activation('sigmoid'))
five_layer_model.add(Dropout(0.10))

# Output
five_layer_model.add(Dense(1, kernel_initializer='uniform')) 
five_layer_model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
five_layer_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

five_layer_model.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test,five_layer_model.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 24s 16us/step - loss: 0.6923 - acc: 0.5243
Epoch 2/5
1500000/1500000 [==============================] - 22s 15us/step - loss: 0.6921 - acc: 0.5257
Epoch 3/5
1500000/1500000 [==============================] - 21s 14us/step - loss: 0.6919 - acc: 0.5275
Epoch 4/5
1500000/1500000 [==============================] - 24s 16us/step - loss: 0.6918 - acc: 0.5283
Epoch 5/5
1500000/1500000 [==============================] - 22s 15us/step - loss: 0.6917 - acc: 0.5287


0.4953698768841075

In [32]:
#### Model 3 - 1 hidden layers and varying nodes, only 1 drop out

In [35]:
model3 = Sequential()

# Input Layer
model3.add(Dense(25, input_dim = x.shape[1], kernel_initializer='uniform'))
model3.add(Activation('sigmoid'))

# First Hidden Layer
model3.add(Dense(50, kernel_initializer='uniform'))
model3.add(Activation('sigmoid'))
model3.add(Dropout(0.1))

# output
model3.add(Dense(1, kernel_initializer='uniform'))
model3.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model3.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model3.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, model3.predict(x_test))


Epoch 1/5
1500000/1500000 [==============================] - 9s 6us/step - loss: 0.6920 - acc: 0.5270
Epoch 2/5
1500000/1500000 [==============================] - 8s 5us/step - loss: 0.6919 - acc: 0.5275
Epoch 3/5
1500000/1500000 [==============================] - 8s 5us/step - loss: 0.6907 - acc: 0.5312
Epoch 4/5
1500000/1500000 [==============================] - 8s 6us/step - loss: 0.6543 - acc: 0.6085
Epoch 5/5
1500000/1500000 [==============================] - 8s 5us/step - loss: 0.6357 - acc: 0.6339


0.705244511974489

#### Model 4 - 1 hidden layer and lots of nodes

In [36]:
model4 = Sequential()

# Input Layer
model4.add(Dense(100, input_dim = x.shape[1], kernel_initializer='uniform'))
model4.add(Activation('sigmoid'))

# First Hidden Layer
model4.add(Dense(200, kernel_initializer='uniform'))
model4.add(Activation('sigmoid'))
model4.add(Dropout(0.1))

# output
model4.add(Dense(1, kernel_initializer='uniform'))
model4.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model4.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model4.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, model4.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 14s 10us/step - loss: 0.6929 - acc: 0.5234
Epoch 2/5
1500000/1500000 [==============================] - 13s 9us/step - loss: 0.6523 - acc: 0.6111
Epoch 3/5
1500000/1500000 [==============================] - 13s 9us/step - loss: 0.6406 - acc: 0.6287
Epoch 4/5
1500000/1500000 [==============================] - 14s 9us/step - loss: 0.6318 - acc: 0.6412
Epoch 5/5
1500000/1500000 [==============================] - 13s 9us/step - loss: 0.6202 - acc: 0.6557


0.7256707343472436

### Part 2
With those same 3 architectures, run the SAME architecture but with 2 different (from sigmoid) activation functions.  Google the Keras documentation for a look at different available activations. 

#### Activation = 'softmax'
Five Layer Model - No Improvement 

Model 3 - Decrease in accuracy 

Model 4 - Major decrease in accuracy 

In [49]:
five_layer_model_softmax = Sequential()
# Layer 1
five_layer_model_softmax.add(Dense(50, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 28 here
five_layer_model_softmax.add(Activation('softmax'))
five_layer_model_softmax.add(Dropout(0.10))

# Layer 2
five_layer_model_softmax.add(Dense(50, kernel_initializer='uniform'))
five_layer_model_softmax.add(Activation('softmax'))
five_layer_model_softmax.add(Dropout(0.10))

# Layer 3
five_layer_model_softmax.add(Dense(50, kernel_initializer='uniform')) 
five_layer_model_softmax.add(Activation('softmax'))
five_layer_model_softmax.add(Dropout(0.10))

# Layer 4
five_layer_model_softmax.add(Dense(50, kernel_initializer='uniform')) 
five_layer_model_softmax.add(Activation('softmax'))
five_layer_model_softmax.add(Dropout(0.10))

# Layer 5
five_layer_model_softmax.add(Dense(50, kernel_initializer='uniform')) 
five_layer_model_softmax.add(Activation('softmax'))
five_layer_model_softmax.add(Dropout(0.10))

# Output
five_layer_model_softmax.add(Dense(1, kernel_initializer='uniform')) 
five_layer_model_softmax.add(Activation('softmax'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
five_layer_model_softmax.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

five_layer_model_softmax.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test,five_layer_model_softmax.predict(x_test))


Epoch 1/5
1500000/1500000 [==============================] - 19s 13us/step - loss: 7.5060 - acc: 0.5292
Epoch 2/5
1500000/1500000 [==============================] - 18s 12us/step - loss: 7.5060 - acc: 0.5292
Epoch 3/5
1500000/1500000 [==============================] - 18s 12us/step - loss: 7.5060 - acc: 0.5292
Epoch 4/5
1500000/1500000 [==============================] - 18s 12us/step - loss: 7.5060 - acc: 0.5292
Epoch 5/5
1500000/1500000 [==============================] - 18s 12us/step - loss: 7.5060 - acc: 0.5292


0.5

In [50]:
model3_softmax = Sequential()

# Input Layer
model3_softmax.add(Dense(25, input_dim = x.shape[1], kernel_initializer='uniform'))
model3_softmax.add(Activation('softmax'))

# First Hidden Layer
model3_softmax.add(Dense(50, kernel_initializer='uniform'))
model3_softmax.add(Activation('softmax'))
model3_softmax.add(Dropout(0.1))

# output
model3_softmax.add(Dense(1, kernel_initializer='uniform'))
model3_softmax.add(Activation('softmax'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model3_softmax.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model3_softmax.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, model3_softmax.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 11s 7us/step - loss: 7.5060 - acc: 0.5292
Epoch 2/5
1500000/1500000 [==============================] - 9s 6us/step - loss: 7.5060 - acc: 0.5292
Epoch 3/5
1500000/1500000 [==============================] - 9s 6us/step - loss: 7.5060 - acc: 0.5292
Epoch 4/5
1500000/1500000 [==============================] - 9s 6us/step - loss: 7.5060 - acc: 0.5292
Epoch 5/5
1500000/1500000 [==============================] - 10s 6us/step - loss: 7.5060 - acc: 0.5292


0.5

In [51]:
model4_softmax = Sequential()

# Input Layer
model4_softmax.add(Dense(100, input_dim = x.shape[1], kernel_initializer='uniform'))
model4_softmax.add(Activation('softmax'))

# First Hidden Layer
model4_softmax.add(Dense(200, kernel_initializer='uniform'))
model4_softmax.add(Activation('softmax'))
model4_softmax.add(Dropout(0.1))

# output
model4_softmax.add(Dense(1, kernel_initializer='uniform'))
model4_softmax.add(Activation('softmax'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model4_softmax.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model4_softmax.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, model4_softmax.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 17s 12us/step - loss: 7.5060 - acc: 0.5292
Epoch 2/5
1500000/1500000 [==============================] - 16s 11us/step - loss: 7.5060 - acc: 0.5292
Epoch 3/5
1500000/1500000 [==============================] - 16s 11us/step - loss: 7.5060 - acc: 0.5292
Epoch 4/5
1500000/1500000 [==============================] - 16s 11us/step - loss: 7.5060 - acc: 0.5292
Epoch 5/5
1500000/1500000 [==============================] - 16s 11us/step - loss: 7.5060 - acc: 0.5292


0.5

# Activation Function = RELU
Five Layer Model - Major Decrease in Accuracy. Only predicts the major class.

Model3 - Major Decrease in Accuracy. Only predicts the major class.

Model4 - Major Decrase in Accuracy. Only predicts the major class.

In [46]:
model3_relu = Sequential()

# Input Layer
model3_relu.add(Dense(25, input_dim = x.shape[1], kernel_initializer='uniform'))
model3_relu.add(Activation('relu'))

# First Hidden Layer
model3_relu.add(Dense(50, kernel_initializer='uniform'))
model3_relu.add(Activation('relu'))
model3_relu.add(Dropout(0.1))

# output
model3_relu.add(Dense(1, kernel_initializer='uniform'))
model3_relu.add(Activation('relu'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model3_relu.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model3_relu.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, model3_relu.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 10s 7us/step - loss: 7.5064 - acc: 3.0933e-04
Epoch 2/5
1500000/1500000 [==============================] - 8s 6us/step - loss: 7.5060 - acc: 0.0000e+00
Epoch 3/5
1500000/1500000 [==============================] - 8s 6us/step - loss: 7.5060 - acc: 0.0000e+00
Epoch 4/5
1500000/1500000 [==============================] - 8s 6us/step - loss: 7.5060 - acc: 0.0000e+00
Epoch 5/5
1500000/1500000 [==============================] - 8s 6us/step - loss: 7.5060 - acc: 0.0000e+00


0.46727651466928544

In [47]:
model4_relu = Sequential()

# Input Layer
model4_relu.add(Dense(100, input_dim = x.shape[1], kernel_initializer='uniform'))
model4_relu.add(Activation('relu'))

# First Hidden Layer
model4_relu.add(Dense(200, kernel_initializer='uniform'))
model4_relu.add(Activation('relu'))
model4_relu.add(Dropout(0.1))

# output
model4_relu.add(Dense(1, kernel_initializer='uniform'))
model4_relu.add(Activation('relu'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model4_relu.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model4_relu.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, model4_relu.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 16s 10us/step - loss: 7.5028 - acc: 3.1533e-04
Epoch 2/5
1500000/1500000 [==============================] - 13s 9us/step - loss: 7.5060 - acc: 0.0000e+00
Epoch 3/5
1500000/1500000 [==============================] - 14s 9us/step - loss: 7.5060 - acc: 0.0000e+00
Epoch 4/5
1500000/1500000 [==============================] - 14s 9us/step - loss: 7.5060 - acc: 0.0000e+00
Epoch 5/5
1500000/1500000 [==============================] - 14s 9us/step - loss: 7.5060 - acc: 0.0000e+00


0.4785849348100599

#### Mixed Activation
Five layer model - No improvement

Model3 - Improvement

Model4 - Major improvement

In [64]:
five_layer_model_mixed = Sequential()
# Layer 1
five_layer_model_mixed.add(Dense(50, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 28 here
five_layer_model_mixed.add(Activation('relu'))
five_layer_model_mixed.add(Dropout(0.10))

# Layer 2
five_layer_model_mixed.add(Dense(50, kernel_initializer='uniform'))
five_layer_model_mixed.add(Activation('relu'))
five_layer_model_mixed.add(Dropout(0.10))

# Layer 3
five_layer_model_mixed.add(Dense(50, kernel_initializer='uniform')) 
five_layer_model_mixed.add(Activation('relu'))
five_layer_model_mixed.add(Dropout(0.10))

# Layer 4
five_layer_model_mixed.add(Dense(50, kernel_initializer='uniform')) 
five_layer_model_mixed.add(Activation('relu'))
five_layer_model_mixed.add(Dropout(0.10))

# Layer 5
five_layer_model_mixed.add(Dense(50, kernel_initializer='uniform')) 
five_layer_model_mixed.add(Activation('relu'))
five_layer_model_mixed.add(Dropout(0.10))

# Output
five_layer_model_mixed.add(Dense(1, kernel_initializer='uniform')) 
five_layer_model_mixed.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
five_layer_model_mixed.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

five_layer_model_mixed.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test,five_layer_model_mixed.predict(x_test))


Epoch 1/5
1500000/1500000 [==============================] - 19s 12us/step - loss: 0.6915 - acc: 0.5292
Epoch 2/5
1500000/1500000 [==============================] - 16s 11us/step - loss: 0.6915 - acc: 0.5292
Epoch 3/5
1500000/1500000 [==============================] - 16s 11us/step - loss: 0.6915 - acc: 0.5292
Epoch 4/5
1500000/1500000 [==============================] - 16s 11us/step - loss: 0.6915 - acc: 0.5292
Epoch 5/5
1500000/1500000 [==============================] - 17s 11us/step - loss: 0.6915 - acc: 0.5292


0.5639393663769061

In [66]:
model3_softmax = Sequential()

# Input Layer
model3_softmax.add(Dense(25, input_dim = x.shape[1], kernel_initializer='uniform'))
model3_softmax.add(Activation('relu'))

# First Hidden Layer
model3_softmax.add(Dense(50, kernel_initializer='uniform'))
model3_softmax.add(Activation('relu'))
model3_softmax.add(Dropout(0.1))

# output
model3_softmax.add(Dense(1, kernel_initializer='uniform'))
model3_softmax.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model3_softmax.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model3_softmax.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, model3_softmax.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 11s 7us/step - loss: 0.6145 - acc: 0.6514
Epoch 2/5
1500000/1500000 [==============================] - 9s 6us/step - loss: 0.5663 - acc: 0.7037
Epoch 3/5
1500000/1500000 [==============================] - 9s 6us/step - loss: 0.5531 - acc: 0.7139
Epoch 4/5
1500000/1500000 [==============================] - 9s 6us/step - loss: 0.5447 - acc: 0.7200
Epoch 5/5
1500000/1500000 [==============================] - 9s 6us/step - loss: 0.5379 - acc: 0.7247


0.8082438014584612

In [63]:
model4_sigmoid = Sequential()

# Input Layer
model4_sigmoid.add(Dense(100, input_dim = x.shape[1], kernel_initializer='uniform'))
model4_sigmoid.add(Activation('relu'))

# First Hidden Layer
model4_sigmoid.add(Dense(200, kernel_initializer='uniform'))
model4_sigmoid.add(Activation('relu'))
model4_sigmoid.add(Dropout(0.1))

# output
model4_sigmoid.add(Dense(1, kernel_initializer='uniform'))
model4_sigmoid.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model4_sigmoid.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model4_sigmoid.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, model4_sigmoid.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5990 - acc: 0.6701
Epoch 2/5
1500000/1500000 [==============================] - 14s 10us/step - loss: 0.5556 - acc: 0.7108
Epoch 3/5
1500000/1500000 [==============================] - 14s 10us/step - loss: 0.5388 - acc: 0.7239
Epoch 4/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5274 - acc: 0.7325
Epoch 5/5
1500000/1500000 [==============================] - 14s 9us/step - loss: 0.5205 - acc: 0.7373


0.8205968128607934

In [69]:
model4_sigmoid = Sequential()

# Input Layer
model4_sigmoid.add(Dense(100, input_dim = x.shape[1], kernel_initializer='uniform'))
model4_sigmoid.add(Activation('tanh'))

# First Hidden Layer
model4_sigmoid.add(Dense(200, kernel_initializer='uniform'))
model4_sigmoid.add(Activation('tanh'))
model4_sigmoid.add(Dropout(0.1))

# output
model4_sigmoid.add(Dense(1, kernel_initializer='uniform'))
model4_sigmoid.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model4_sigmoid.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model4_sigmoid.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, model4_sigmoid.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 16s 11us/step - loss: 0.6263 - acc: 0.6473
Epoch 2/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5890 - acc: 0.6876
Epoch 3/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5761 - acc: 0.6987
Epoch 4/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5695 - acc: 0.7029
Epoch 5/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5649 - acc: 0.7061


0.7806258597860792

### Part 3
Take your best model from parts 1&2 and vary the batch size by at least 2 orders of magnitude

#### Batch sizes of 100 and 10,000

In [71]:
best_model = Sequential()

# Input Layer
best_model.add(Dense(100, input_dim = x.shape[1], kernel_initializer='uniform'))
best_model.add(Activation('relu'))

# First Hidden Layer
best_model.add(Dense(200, kernel_initializer='uniform'))
best_model.add(Activation('relu'))
best_model.add(Dropout(0.1))

# output
best_model.add(Dense(1, kernel_initializer='uniform'))
best_model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
best_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

best_model.fit(x, y, epochs=5, batch_size=100)
roc_auc_score(y_test, best_model.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 57s 38us/step - loss: 0.5882 - acc: 0.6845
Epoch 2/5
1500000/1500000 [==============================] - 54s 36us/step - loss: 0.5602 - acc: 0.7098
Epoch 3/5
1500000/1500000 [==============================] - 55s 37us/step - loss: 0.5485 - acc: 0.7195
Epoch 4/5
1500000/1500000 [==============================] - 54s 36us/step - loss: 0.5417 - acc: 0.7242
Epoch 5/5
1500000/1500000 [==============================] - 54s 36us/step - loss: 0.5377 - acc: 0.7268


0.8088964609753136

In [72]:
best_model = Sequential()

# Input Layer
best_model.add(Dense(100, input_dim = x.shape[1], kernel_initializer='uniform'))
best_model.add(Activation('relu'))

# First Hidden Layer
best_model.add(Dense(200, kernel_initializer='uniform'))
best_model.add(Activation('relu'))
best_model.add(Dropout(0.1))

# output
best_model.add(Dense(1, kernel_initializer='uniform'))
best_model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
best_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

best_model.fit(x, y, epochs=5, batch_size=10000)
roc_auc_score(y_test, best_model.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 13s 9us/step - loss: 0.6743 - acc: 0.5710
Epoch 2/5
1500000/1500000 [==============================] - 10s 7us/step - loss: 0.6274 - acc: 0.6458
Epoch 3/5
1500000/1500000 [==============================] - 10s 7us/step - loss: 0.6090 - acc: 0.6654
Epoch 4/5
1500000/1500000 [==============================] - 10s 7us/step - loss: 0.5959 - acc: 0.6789
Epoch 5/5
1500000/1500000 [==============================] - 10s 7us/step - loss: 0.5868 - acc: 0.6869


0.7657946292650923

#### Part 4
Take your best model (score) from parts 1&2 and use 3 different kernel initializers. Use a reasonable batch size.

Some kernel initializers decreased performance. I didn't find any that greatly improved performance.

In [73]:
# Zeros
best_model = Sequential()

# Input Layer
best_model.add(Dense(100, input_dim = x.shape[1], kernel_initializer='zeros'))
best_model.add(Activation('relu'))

# First Hidden Layer
best_model.add(Dense(200, kernel_initializer='zeros'))
best_model.add(Activation('relu'))
best_model.add(Dropout(0.1))

# output
best_model.add(Dense(1, kernel_initializer='zeros'))
best_model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
best_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

best_model.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, best_model.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 17s 11us/step - loss: 0.6915 - acc: 0.5291
Epoch 2/5
1500000/1500000 [==============================] - 14s 10us/step - loss: 0.6915 - acc: 0.5292
Epoch 3/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.6915 - acc: 0.5292
Epoch 4/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.6915 - acc: 0.5292
Epoch 5/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.6915 - acc: 0.5292


0.5

In [74]:
# Normal
best_model = Sequential()

# Input Layer
best_model.add(Dense(100, input_dim = x.shape[1], kernel_initializer='normal'))
best_model.add(Activation('relu'))

# First Hidden Layer
best_model.add(Dense(200, kernel_initializer='normal'))
best_model.add(Activation('relu'))
best_model.add(Dropout(0.1))

# output
best_model.add(Dense(1, kernel_initializer='normal'))
best_model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
best_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

best_model.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, best_model.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 17s 12us/step - loss: 0.5967 - acc: 0.6734
Epoch 2/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5555 - acc: 0.7115
Epoch 3/5
1500000/1500000 [==============================] - 14s 10us/step - loss: 0.5384 - acc: 0.7243
Epoch 4/5
1500000/1500000 [==============================] - 14s 10us/step - loss: 0.5268 - acc: 0.7324
Epoch 5/5
1500000/1500000 [==============================] - 14s 9us/step - loss: 0.5196 - acc: 0.7375


0.8209072021446804

In [75]:
# glorot_uniform
# glorot_uniform
best_model = Sequential()

# Input Layer
best_model.add(Dense(100, input_dim = x.shape[1], kernel_initializer='glorot_uniform'))
best_model.add(Activation('relu'))

# First Hidden Layer
best_model.add(Dense(200, kernel_initializer='glorot_uniform'))
best_model.add(Activation('relu'))
best_model.add(Dropout(0.1))

# output
best_model.add(Dense(1, kernel_initializer='glorot_uniform'))
best_model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
best_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

best_model.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, best_model.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 17s 11us/step - loss: 0.5978 - acc: 0.6742
Epoch 2/5
1500000/1500000 [==============================] - 14s 9us/step - loss: 0.5580 - acc: 0.7104
Epoch 3/5
1500000/1500000 [==============================] - 14s 9us/step - loss: 0.5413 - acc: 0.7233
Epoch 4/5
1500000/1500000 [==============================] - 14s 10us/step - loss: 0.5310 - acc: 0.7308
Epoch 5/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5243 - acc: 0.7353


0.8190581645863504

In [76]:
# lecon_lecun_normal
# lecun_normal
best_model = Sequential()

# Input Layer
best_model.add(Dense(100, input_dim = x.shape[1], kernel_initializer='lecun_normal'))
best_model.add(Activation('relu'))

# First Hidden Layer
best_model.add(Dense(200, kernel_initializer='lecun_normal'))
best_model.add(Activation('relu'))
best_model.add(Dropout(0.1))

# output
best_model.add(Dense(1, kernel_initializer='lecun_normal'))
best_model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
best_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

best_model.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, best_model.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 17s 12us/step - loss: 0.6011 - acc: 0.6716
Epoch 2/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5594 - acc: 0.7088
Epoch 3/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5420 - acc: 0.7221
Epoch 4/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5317 - acc: 0.7298
Epoch 5/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5250 - acc: 0.7346


0.8201600881403961

In [77]:
best_model = Sequential()

# Input Layer
best_model.add(Dense(100, input_dim = x.shape[1], kernel_initializer='VarianceScaling'))
best_model.add(Activation('relu'))

# First Hidden Layer
best_model.add(Dense(200, kernel_initializer='VarianceScaling'))
best_model.add(Activation('relu'))
best_model.add(Dropout(0.1))

# output
best_model.add(Dense(1, kernel_initializer='VarianceScaling'))
best_model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
best_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

best_model.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, best_model.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 17s 11us/step - loss: 0.6018 - acc: 0.6707
Epoch 2/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5624 - acc: 0.7074
Epoch 3/5
1500000/1500000 [==============================] - 14s 10us/step - loss: 0.5456 - acc: 0.7205
Epoch 4/5
1500000/1500000 [==============================] - 14s 10us/step - loss: 0.5351 - acc: 0.7281
Epoch 5/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5274 - acc: 0.7332


0.8185766944464924

#### Part 5 - Optimizers
Take your best results from #3 and try 3 different optimizers.

In [80]:
# Normal
best_model = Sequential()

# Input Layer
best_model.add(Dense(100, input_dim = x.shape[1], kernel_initializer='normal'))
best_model.add(Activation('relu'))

# First Hidden Layer
best_model.add(Dense(200, kernel_initializer='normal'))
best_model.add(Activation('relu'))
best_model.add(Dropout(0.1))

# output
best_model.add(Dense(1, kernel_initializer='normal'))
best_model.add(Activation('sigmoid'))

rms = RMSprop(lr=0.001, rho=0.9, epsilon=None, decay=0.0)
best_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=rms)

best_model.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, best_model.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 18s 12us/step - loss: 0.6139 - acc: 0.6578
Epoch 2/5
1500000/1500000 [==============================] - 16s 10us/step - loss: 0.5736 - acc: 0.6973
Epoch 3/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5574 - acc: 0.7112
Epoch 4/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5466 - acc: 0.7193
Epoch 5/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5379 - acc: 0.7257


0.8075944749806508

In [84]:
# Normal
best_model = Sequential()

# Input Layer
best_model.add(Dense(100, input_dim = x.shape[1], kernel_initializer='normal'))
best_model.add(Activation('relu'))

# First Hidden Layer
best_model.add(Dense(200, kernel_initializer='normal'))
best_model.add(Activation('relu'))
best_model.add(Dropout(0.1))

# output
best_model.add(Dense(1, kernel_initializer='normal'))
best_model.add(Activation('sigmoid'))

ada = Adadelta(lr=1.0, rho=0.95, epsilon=None, decay=0.0)
best_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=ada)

best_model.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, best_model.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 18s 12us/step - loss: 0.6156 - acc: 0.6545
Epoch 2/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5740 - acc: 0.6959
Epoch 3/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5578 - acc: 0.7097
Epoch 4/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5472 - acc: 0.7179
Epoch 5/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5385 - acc: 0.7245


0.8079452554585951

In [85]:
# Normal
best_model = Sequential()

# Input Layer
best_model.add(Dense(100, input_dim = x.shape[1], kernel_initializer='normal'))
best_model.add(Activation('relu'))

# First Hidden Layer
best_model.add(Dense(200, kernel_initializer='normal'))
best_model.add(Activation('relu'))
best_model.add(Dropout(0.1))

# output
best_model.add(Dense(1, kernel_initializer='normal'))
best_model.add(Activation('sigmoid'))

adm = Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0, amsgrad=False)
best_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=adm)

best_model.fit(x, y, epochs=5, batch_size=1000)
roc_auc_score(y_test, best_model.predict(x_test))

Epoch 1/5
1500000/1500000 [==============================] - 18s 12us/step - loss: 0.5994 - acc: 0.6728
Epoch 2/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5630 - acc: 0.7068
Epoch 3/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5472 - acc: 0.7187
Epoch 4/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5359 - acc: 0.7270
Epoch 5/5
1500000/1500000 [==============================] - 15s 10us/step - loss: 0.5274 - acc: 0.7331


0.8168511284648247

#### Part 6:
Take all that you’ve learned so far and give your best shot at producing a score. 

Best model so far is below with an acc of 0.7373 and a ROC of 0.82

In [4]:
# Normal
best_model = Sequential()

# Input Layer
best_model.add(Dense(200, input_dim = x.shape[1], kernel_initializer='normal'))
best_model.add(Activation('relu'))

# First Hidden Layer
best_model.add(Dense(300, kernel_initializer='normal'))
best_model.add(Activation('relu'))
best_model.add(Dropout(0.1))

# output
best_model.add(Dense(1, kernel_initializer='normal'))
best_model.add(Activation('sigmoid'))

sgd = SGD(lr=0.12, decay=1e-6, momentum=0.9, nesterov=True)
best_model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

best_model.fit(x, y, epochs=5, batch_size=10000)
roc_auc_score(y_test, best_model.predict(x_test))

Epoch 1/5
 680000/1500000 [============>.................] - ETA: 15s - loss: 0.6718 - acc: 0.5834

KeyboardInterrupt: 

In [6]:
x.shape

(1500000, 28)